### ISSUE TITLE
match_same_arms should ignore todo!() branches

### ISSUE THREAD

**Issue by Gaelan** [rust-lang/rust-clippy#7331]
Lint name:
match_same_arms

I tried [this code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a07b322a2f140d8d42e008bc8282c5d7):

```rust
#![warn(clippy::pedantic)]
#![allow(dead_code, clippy::blacklisted_name)]

#[derive(Copy, Clone)]
enum Foo {A, B}

fn test(foo: Foo) {
    match foo {
        Foo::A => todo!(),
        Foo::B => todo!()
    }
}
```

I expected to see this happen:

No warnings. In this case, it's pretty clear that, eventually, we intend to do different things for A and B; combining them would only create more work when we inevitably split them apart later.

It's also worth noting that this structure (several match arms containing `todo!()`) is generated by rust-analyzer's "fill match arms" feature.

Instead, this happened:

Clippy warns me:

```
warning: this `match` has identical arm bodies
  --> src/lib.rs:10:19
   |
10 |         Foo::B => todo!()
   |                   ^^^^^^^
   |
note: the lint level is defined here
  --> src/lib.rs:1:9
   |
1  | #![warn(clippy::pedantic)]
   |         ^^^^^^^^^^^^^^^^
   = note: `#[warn(clippy::match_same_arms)]` implied by `#[warn(clippy::pedantic)]`
note: same as this
  --> src/lib.rs:9:19
   |
9  |         Foo::A => todo!(),
   |                   ^^^^^^^
help: consider refactoring into `Foo::A | Foo::B`
  --> src/lib.rs:9:9
   |
9  |         Foo::A => todo!(),
   |         ^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_same_arms
   = note: this warning originates in the macro `$crate::panic::panic_2015` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Meta

Note: the code in this issue was tested in the playground, which I can't figure out how to get version information for; but I had the same issue (with real code) locally, with the below versions.

- `cargo clippy -V`: clippy 0.1.52 (07e0e2ec 2021-03-24)
- `rustc -Vv`:
  ```
  rustc 1.53.0-nightly (07e0e2ec2 2021-03-24)
  binary: rustc
  commit-hash: 07e0e2ec268c140e607e1ac7f49f145612d0f597
  commit-date: 2021-03-24
  host: x86_64-apple-darwin
  release: 1.53.0-nightly
  LLVM version: 12.0.0
  ```

**Comment by flip1995** [comment]
I mean the lint is correct, that you don't want to have two match arms with the same body. I don't see this as a FP. Once you fill those arms the lint will stop triggering. If, for some reason, you can't implement this yet, but need Clippy to pass anyway, you should add an `allow`. 

So you're saying, you still have to change this match arms and Clippy tells you that those match arms are the same and you should do something about it. So I would say Clippy does the right thing here.

Regarding the suggestion: This isn't what you want to do in this case. But we can't really tell what the original intention was. Adding an explicit check for `todo!` seems a bit excessive to me. We may want to add a general message, that another possible fix would be to change the match arm body, instead of combining the match arms.

Also this lint is in the `pedantic` group, so I would say being pedantic here with the `todo!` macros is fine.
